tokendirectory.go

```diff
@@ -70,195 +73,152 @@ type TokenDirectory struct {
 
 type OnUpdateFunc func(ctx context.Context, chainID uint64, contractInfoList []ContractInfo)
 
-// SetHttpClient sets the http client used to fetch token-lists from remote sources.
-func (f *TokenDirectory) SetHttpClient(client *http.Client) {
-	f.httpClient = client
-}
-
 // Run starts the token directory fetcher. This method will block and poll in the current
 // go-routine. You'll be responsible for calling the Run method in your own gorutine.
-func (f *TokenDirectory) Run(ctx context.Context) error {
-	if f.IsRunning() {
+func (t *TokenDirectory) Run(ctx context.Context) error {
+	if t.IsRunning() {
 		return fmt.Errorf("tokendirectory: already running")
 	}
 
-	f.ctx, f.ctxStop = context.WithCancel(ctx)
+	t.ctx, t.ctxStop = context.WithCancel(ctx)
 
-	atomic.StoreInt32(&f.running, 1)
-	defer atomic.StoreInt32(&f.running, 0)
+	atomic.StoreInt32(&t.running, 1)
+	defer atomic.StoreInt32(&t.running, 0)
 
 	// Initial source fetch
-	f.updateSources(f.ctx)
+	t.updateSources(t.ctx)
 
 	// Fetch on interval
 	for {
 		select {
-		case <-f.ctx.Done():
+		case <-t.ctx.Done():
 			return nil
-		case <-time.After(f.updateInterval):
-			f.updateSources(f.ctx)
+		case <-time.After(t.updateInterval):
+			t.updateSources(t.ctx)
 		}
 	}
 }
 
-func (f *TokenDirectory) Stop() {
+func (t *TokenDirectory) Stop() {
 	log.Info().Msgf("tokendirectory: stop")
-	f.ctxStop()
+	t.ctxStop()
 }
 
-func (f *TokenDirectory) IsRunning() bool {
-	return atomic.LoadInt32(&f.running) == 1
+func (t *TokenDirectory) IsRunning() bool {
+	return atomic.LoadInt32(&t.running) == 1
 }
 
-func (f *TokenDirectory) updateSources(ctx context.Context) error {
+func (t *TokenDirectory) updateSources(ctx context.Context) error {
 	wg := &sync.WaitGroup{}
-	for _, source := range f.sources {
-		for _, chainID := range source.GetChainIDs() {
-			wg.Add(1)
-			go func(chainID uint64, source Source) {
-				defer wg.Done()
-				f.updateChainSource(ctx, source, chainID)
-			}(chainID, source)
+	for _, provider := range t.providers {
+		for _, chainID := range provider.GetChainIDs() {
+			for _, source := range provider.GetSources(chainID) {
+				wg.Add(1)
+				go func(provider Provider, chainID uint64, source string) {
+					defer wg.Done()
+					t.updateProvider(ctx, provider, chainID, source)
+				}(provider, chainID, source)
+			}
 		}
 	}
 	wg.Wait()
 	return nil
 }
 
-func (f *TokenDirectory) updateChainSource(ctx context.Context, source Source, chainID uint64) {
-	f.updateMu.Lock()
-	defer f.updateMu.Unlock()
+func (t *TokenDirectory) updateProvider(ctx context.Context, provider Provider, chainID uint64, source string) {
+	t.updateMu.Lock()
```

Comment:
This way, the logging can be done on one place without nil checks for logger.
```suggestion
	var err error
	defer func() {
		if t.log != nil {
			logger := t.log.With(
				slog.String("providerID", provider.GetID()),
				slog.Uint64("chainID", chainID),
				slog.String("source", source),
			)

			logger.Debug("fetching token list")
			if err != nil {
				logger.Error("failed to fetch token list", slog.Any("err", err))
			}
		}

		t.updateMu.Unlock()
	}()

	t.updateMu.Lock()

	updatedContractInfo := []ContractInfo{}
	seen := map[string]bool{}

	t.contractsMu.Lock()
	if _, ok := t.lists[chainID]; !ok {
		t.lists[chainID] = make(map[string]*TokenList)
	}
	if _, ok := t.contracts[chainID]; !ok {
		t.contracts[chainID] = make(map[prototyp.Hash]ContractInfo)
	}
	t.contractsMu.Unlock()

	tokenList, err := provider.FetchTokenList(ctx, chainID, source)
	if err != nil {
		return
	}

	normalizeTokens(provider, tokenList)

	t.lists[chainID][provider.GetID()] = tokenList

```